runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java

```diff
@@ -997,6 +1003,90 @@ private static void translateFn(
     }
   }
 
+  /**
+   * Builds a {@link RunnerApi.PTransform} with adequate context to supply all the information
+   * needed for the Dataflow worker to execute the {@link DoFn} without having to inspect a larger
+   * subgraph.
+   */
+  private static byte[] payloadForFn(
+      StepTranslationContext stepContext,
+      DoFn fn,
+      PCollection<?> input,
+      Iterable<PCollectionView<?>> sideInputViews,
+      Coder inputCoder,
+      TupleTag<?> mainOutputTag,
+      Map<TupleTag<?>, PValue> outputs)
+      throws IOException {
+
+    // The root message to embed in the Dataflow API layer
+    RunnerApi.MessageWithComponents.Builder message = RunnerApi.MessageWithComponents.newBuilder();
+
+    // The surrounding graph context needed to re-wire things on the worker
+    // We need to know the main input, side inputs, all outputs, and distinguished main output
+    SdkComponents sdkComponents = SdkComponents.create();
+    RunnerApi.PTransform.Builder protoTransform = RunnerApi.PTransform.newBuilder();
+
+    // The ParDo-specific payload
+    RunnerApi.ParDoPayload.Builder parDoPayload = RunnerApi.ParDoPayload.newBuilder();
+
+    // The main input tag can be arbitrary, as long as it does not match any side input tag
```

Comment:
<!--thread_id:cc_139777656_t; commit:c619b6e149fe93f1d533b7b7fe6c5d4bd1a7fdb3; resolved:1-->
<!--section:context-quote-->
> **tgroh** wrote:
> It is theoretically (though not via any of our provided Side-input creation APIs) possible to create a view with this tag. Is it worth validating? We could do so in the iteration over the views, or add the main input last (and throw if it collides)

<!--section:body-->
I realized that expand() provides a tag, with whatever guarantees are necessary, presumably?